### PR TITLE
Remove broken is_on_device macro and its use in SparseVector.

### PR DESCRIPTION
**Pull request by agheata**
This removes the use of the is_on_device macro that confused the compiler producing undetermined behavior.

### CODE REVIEW

File: base/inc/CopCore/include/CopCore/Launcher.h [hunk 1]
```diff
@@ -99,8 +99,8 @@ class Launcher<BackendType::CUDA> : public LauncherBase<BackendType::CUDA> {
           std::min(warpsPerSM * fNumSMs * 32 / block_size, (n_elements + block_size - 1) / block_size);
       exec_grid[0].x = grid_size;
       exec_grid[1].x = block_size;
-      // std::cout << "grid_size = " << grid_size << "  block_size = " << block_size << std::endl;
     }
+    //std::cout << "grid_size = " << exec_grid[0].x << "  block_size = " << exec_grid[1].x << std::endl;
```
Comment (hahnjo) [inline base/inc/CopCore/include/CopCore/Launcher.h:103]: unrelated

**Review by hahnjo — CHANGES_REQUESTED** [review]

File: base/inc/AdePT/SparseVector.h [hunk 2]
```diff
@@ -718,12 +718,14 @@ class SparseVector : public SparseVectorInterface<Type> {
...
   __host__ static SparseVector *MakeInstanceAt(void *addr = nullptr)
   {
     if (!addr) return new SparseVector(N);
-    bool devicePtr = copcore::is_device_pointer(addr);
-    if (devicePtr) return SparseVector::ConstructOnDevice(addr);
-    return new (addr) SparseVector(N);
+    if (backend == copcore::BackendType::CUDA)
+      return SparseVector::ConstructOnDevice(addr);
+    else
+      return new (addr) SparseVector(N);
```
[the remark this answers is not in the document]
Comment (agheata) [reply base/inc/AdePT/SparseVector.h]: Partially specializing the class method is tricky (at least I did not find an obvious way to do it), but I made the implementation cleaner for the CPU case. The allocation with new is currently needed for the CPU case.

File: base/inc/CopCore/include/CopCore/Launcher.h (hunk 1 above)
Comment (hahnjo) [reply base/inc/CopCore/include/CopCore/Launcher.h]: still unrelated

File: examples/Raytracer_Benchmark/RaytraceBenchmark.cu [hunk 3]
```diff
@@ -24,6 +24,13 @@
 #include <cstdio>
 #include <vector>
 
+__global__ void LocateViewpointKernel(cuda::RaytracerData_t *rtdata)
+{
+  // Initialize the navigation state for the view point
+  vecgeom::NavStateIndex vpstate;
+  LoopNavigator::LocatePointIn(rtdata->fWorld, rtdata->fStart, vpstate, true);
+  rtdata->fVPstate = vpstate;
+}
```
[the remark this answers is not in the document]
Comment (agheata) [reply examples/Raytracer_Benchmark/RaytraceBenchmark.cu]: This is going to enter in the PR that changes the RaytracerBenchmark more deeply.

File: examples/Raytracer_Benchmark/RaytraceBenchmark.hpp [hunk 4]
```diff
@@ -161,6 +159,9 @@ int runSimulation(const MyMediumProp *volume_container, const vecgeom::cxx::VPla
   // Use reflection
   OPTION_BOOL(reflection, 0);
 
+  // Approach solids first
+  OPTION_BOOL(approach, 0);
+
```
Comment (hahnjo) [inline examples/Raytracer_Benchmark/RaytraceBenchmark.hpp:164]: Should not be in this PR

File: examples/Raytracer_Benchmark/RaytraceBenchmark.hpp [hunk 5]
```diff
@@ -323,7 +324,8 @@ int runSimulation(const MyMediumProp *volume_container, const vecgeom::cxx::VPla
   COPCORE_CUDA_CHECK(cudaFree(color));
   COPCORE_CUDA_CHECK(cudaFree(sel_vector_d));
   COPCORE_CUDA_CHECK(cudaFree(nselected_hd));
-  COPCORE_CUDA_CHECK(cudaFree(rtdata));
+  rayAlloc.deallocate(rtdata, 1);
+  // COPCORE_CUDA_CHECK(cudaFree(rtdata));
```
Comment (hahnjo) [inline examples/Raytracer_Benchmark/RaytraceBenchmark.hpp:328]: I think this shouldn't be in this PR either

File: base/inc/CopCore/include/CopCore/Launcher.h (hunk 1 above)
Comment (agheata) [reply base/inc/CopCore/include/CopCore/Launcher.h]: fixed now

File: examples/Raytracer_Benchmark/RaytraceBenchmark.hpp (hunk 4 above)
Comment (agheata) [reply examples/Raytracer_Benchmark/RaytraceBenchmark.hpp]: fixed

File: examples/Raytracer_Benchmark/RaytraceBenchmark.hpp (hunk 5 above)
Comment (agheata) [reply examples/Raytracer_Benchmark/RaytraceBenchmark.hpp]: sure, thanks for noticing, these are all leftovers from splitting from #137

**Review by hahnjo — APPROVED** [review]
LGTM (the whitespace changes make the diff hard to read, but eventually I found GitHub's option to hide them...)
